### DISCUSSION TITLE
Zsh-Powerlevel10k

### DISCUSSION THREAD

**Discussion by HFrancisla** [x-cmd/x-cmd#92] [Q&A] [answered]
[WARNING]: Console output during zsh initialization detected.

When using Powerlevel10k with instant prompt, console output during zsh
initialization may indicate issues.

You can:

  - Recommended: Change ~/.zshrc so that it does not perform console I/O
    after the instant prompt preamble. See the link below for details.

    * You will not see this error message again.
    * Zsh will start quickly and prompt will update smoothly.

  - Suppress this warning either by running p10k configure or by manually
    defining the following parameter:

      typeset -g POWERLEVEL9K_INSTANT_PROMPT=quiet

    * You will not see this error message again.
    * Zsh will start quickly but prompt will jump down after initialization.

  - Disable instant prompt either by running p10k configure or by manually
    defining the following parameter:

      typeset -g POWERLEVEL9K_INSTANT_PROMPT=off

    * You will not see this error message again.
    * Zsh will start slowly.

  - Do nothing.

    * You will see this error message every time you start zsh.
    * Zsh will start quickly but prompt will jump down after initialization.

For details, see:
https://github.com/romkatv/powerlevel10k#instant-prompt

-- console output produced during zsh initialization follows --

- I|x: Sorry, we remove the origin x alias to ensure the x-cmd works properly. In the future we will provide a way to choose the short name for ___x_cmd

我想选择这个所谓的recommended做法 ,但我并不清楚具体该如何操作，有人可以帮忙吗QWQ

**Comment by HFrancisla** [comment]
这是我的配置文件 我不太清楚怎么修改 我是小白  
```
# Enable Powerlevel10k instant prompt. Should stay close to the top of ~/.zshrc.
# Initialization code that may require console input (password prompts, [y/n]
# confirmations, etc.) must go above this block; everything else may go below.
if [[ -r "${XDG_CACHE_HOME:-$HOME/.cache}/p10k-instant-prompt-${(%):-%n}.zsh" ]]; then
  source "${XDG_CACHE_HOME:-$HOME/.cache}/p10k-instant-prompt-${(%):-%n}.zsh"
fi

# If you come from bash you might have to change your $PATH.
# export PATH=$HOME/bin:$HOME/.local/bin:/usr/local/bin:$PATH

# Path to your Oh My Zsh installation.
export ZSH="$HOME/.oh-my-zsh"

# Set name of the theme to load --- if set to "random", it will
# load a random theme each time Oh My Zsh is loaded, in which case,
# to know which specific one was loaded, run: echo $RANDOM_THEME
# See https://github.com/ohmyzsh/ohmyzsh/wiki/Themes
ZSH_THEME="powerlevel10k/powerlevel10k"

# Set list of themes to pick from when loading at random
# Setting this variable when ZSH_THEME=random will cause zsh to load
# a theme from this variable instead of looking in $ZSH/themes/
# If set to an empty array, this variable will have no effect.
# ZSH_THEME_RANDOM_CANDIDATES=( "robbyrussell" "agnoster" )

# Uncomment the following line to use case-sensitive completion.
# CASE_SENSITIVE="true"

# Uncomment the following line to use hyphen-insensitive completion.
# Case-sensitive completion must be off. _ and - will be interchangeable.
# HYPHEN_INSENSITIVE="true"

# Uncomment one of the following lines to change the auto-update behavior
# zstyle ':omz:update' mode disabled  # disable automatic updates
# zstyle ':omz:update' mode auto      # update automatically without asking
# zstyle ':omz:update' mode reminder  # just remind me to update when it's time

# Uncomment the following line to change how often to auto-update (in days).
# zstyle ':omz:update' frequency 13

# Uncomment the following line if pasting URLs and other text is messed up.
# DISABLE_MAGIC_FUNCTIONS="true"

# Uncomment the following line to disable colors in ls.
# DISABLE_LS_COLORS="true"
# Uncomment the following line to disable auto-setting terminal title.
# DISABLE_AUTO_TITLE="true"

# Uncomment the following line to enable command auto-correction.
# ENABLE_CORRECTION="true"

# Uncomment the following line to display red dots whilst waiting for completion.
# You can also set it to another string to have that shown instead of the default red dots.
# e.g. COMPLETION_WAITING_DOTS="%F{yellow}waiting...%f"
# Caution: this setting can cause issues with multiline prompts in zsh < 5.7.1 (see #5765)
# COMPLETION_WAITING_DOTS="true"

# Uncomment the following line if you want to disable marking untracked files
# under VCS as dirty. This makes repository status check for large repositories
# much, much faster.
# DISABLE_UNTRACKED_FILES_DIRTY="true"

# Uncomment the following line if you want to change the command execution time
# stamp shown in the history command output.
# You can set one of the optional three formats:
# "mm/dd/yyyy"|"dd.mm.yyyy"|"yyyy-mm-dd"
# or set a custom format using the strftime function format specifications,
# see 'man strftime' for details.
# HIST_STAMPS="mm/dd/yyyy"

# Would you like to use another custom folder than $ZSH/custom?
# ZSH_CUSTOM=/path/to/new-custom-folder

# Which plugins would you like to load?
# Standard plugins can be found in $ZSH/plugins/
# Custom plugins may be added to $ZSH_CUSTOM/plugins/
# Example format: plugins=(rails git textmate ruby lighthouse)
# Add wisely, as too many plugins slow down shell startup.
plugins=(git  extract zsh-syntax-highlighting zsh-autosuggestions )

source $ZSH/oh-my-zsh.sh

# User configuration

[ ! -f "$HOME/.x-cmd.root/X" ] || . "$HOME/.x-cmd.root/X" # boot up x-cmd.

# export MANPATH="/usr/local/man:$MANPATH"

# You may need to manually set your language environment
# export LANG=en_US.UTF-8

# Preferred editor for local and remote sessions
# if [[ -n $SSH_CONNECTION ]]; then
#   export EDITOR='vim'
# else
#   export EDITOR='nvim'
# fi

# Compilation flags
# export ARCHFLAGS="-arch $(uname -m)"

# Set personal aliases, overriding those provided by Oh My Zsh libs,
# plugins, and themes. Aliases can be placed here, though Oh My Zsh
# users are encouraged to define aliases within a top-level file in
# the $ZSH_CUSTOM folder, with .zsh extension. Examples:
# - $ZSH_CUSTOM/aliases.zsh
# - $ZSH_CUSTOM/macos.zsh
# For a full list of active aliases, run `alias`.
#
# Example aliases
# alias zshconfig="mate ~/.zshrc"
# alias ohmyzsh="mate ~/.oh-my-zsh"

# To customize prompt, run `p10k configure` or edit ~/.p10k.zsh.
[[ ! -f ~/.p10k.zsh ]] || source ~/.p10k.zsh

#set man colors
function man {
    env \
        LESS_TERMCAP_mb=$'\E[1;32m' \
        LESS_TERMCAP_md=$'\E[1;32m' \
        LESS_TERMCAP_me=$'\E[0m' \
        LESS_TERMCAP_se=$'\E[0m' \
        LESS_TERMCAP_so=$'\E[01;33m' \
        LESS_TERMCAP_ue=$'\E[0m' \
        LESS_TERMCAP_us=$'\E[1;4;31m' \
            man "$@"
}

export MANPAGER='less -q'

#turn  off the beep
setopt NO_BEEP
unsetopt beep
```

**Comment by edwinjhlee** [comment, marked as answer]
我们没有花时间在 powerlevel10k 这个项目。因为这个项目已经 近乎 deprecated，详情可见 -- [Powerlevel10k](https://github.com/romkatv/powerlevel10k)

我个人建议用 starship 或者 ohmyposh，或者，我们的 theme。

下一个版本 （下周一切换为 latest） 的 x-cmd 会集成 starship 和 ohmyposh，使其更加易于使用。

**Reply by HFrancisla** [reply]
感谢回复，我现在有两个问题

第一个是比如我发现在wsl中使用如x ls 效果就很好，相较之下在git-bash中使用x ls指令，左右切换时屏幕会闪烁地很厉害；

第二个是我想在pwsh上玩这个xcmd，有什么方法吗？  

“虽然我们的产品主要是为了 Linux 和 Unix 体系系统设计，但是，我们也尽力地为 Windows 用户提供不少提高效率的模块。例如与 powershell 的互操作（pwsh 模块），提供了 scoop，winget，choco 模块。”

这一段没看懂什么意思 我应该去哪边找pwsh相关的xcmd？

再表感谢

**Reply by edwinjhlee** [reply]
我们现阶段提供的方案是，在 wsl 或者 git-bash 上调用 pwsh，反过来，在 pwsh 在调用 x-cmd，一直都说要做；但一直都没做（最主要原因可能是，没有人 抱怨，而我们内部的 windows 用户一直都是直接用 wsl 的）。

很可惜是，v0.4.12 刚已生成发布了。我们在下周（ v0.4.13) 将提供 pwsh 调用 x-cmd。这个实现应该难度不高，比较确定性，做法和我们在 其它 posix shell 调用 x-cmd 是一样的。

不过，windows 进程的启动延时一直是一个（似乎不可解决的）大问题。所以在用户体验上，我还是建议 windows 用户：

采用 wsl 的 shell，然后按需调用 pwsh 来处理 windows 环境

我们 后面会进一步增强 pwsh 模块。如果你对这一块感兴趣，欢迎多用，然后就此领域提供意见。

**Reply by edwinjhlee** [reply]
```
第一个是比如我发现在wsl中使用如x ls 效果就很好，相较之下在git-bash中使用x ls指令，左右切换时屏幕会闪烁地很厉害
```

一方面，我们可以对刷新进行更多的优化；但另一方面，git-bash 的实现性能确实是有问题。这些问题很有可能随着未来性能提升而消失，因此我们就没有花时间在这些优化上面。

git-bash 的性能问题还有不少。我们现在 git-bash 的目标是保证其可用，使用体验可接受；但对此有要求的用户，wsl 是最好的选择。所以我们的开发重点是偏向 wsl 的。

另外，后面 x-cmd 提供的一些 安全机制都是基于 docker 和 linux 的。所以，我个人推荐 wsl 作为主力环境。然后把 pwsh 和 git-bash 当作一种部署/功能模块。

**Comment by edwinjhlee** [comment]
我将这个 issue 移动到 discussion。

**Comment by lunrenyi** [comment]
@HzFrancesca 
- 出现警告信息是正常的，在 p10k-instant-prompt 相关的说明中有提及（顶部的注释）。你可以用 `p10k configure` 命令关闭 powerlevel9k 即时提示(INSTANT_PROMPT)带来的信息或者删除 `p10k-instant-prompt` 那三行代码（那是启动powerlevel9k 即时提示的代码）。
- 这个问题在很早以前就有过讨论，你可以到 [Powerlevel10k 社区](https://www.reddit.com/r/zsh/comments/dk53ow/new_powerlevel10k_feature_instant_prompt/) 查找方法
- ohmyposh 的 `plugins=(git  extract zsh-syntax-highlighting zsh-autosuggestions )` 是用来加载插件的，这是导致 zsh 启动变慢的一部分原因。如果你使用 x theme 主题，在 zsh 中，x-cmd 会帮你加载 zsh-syntax-highlighting 和 zsh-autosuggestions 插件。具体你可以查看 https://github.com/x-cmd/x-cmd/discussions/72

ps: 你可以配合 AI 来试着理解你的 .zshrc 配置和 log 的内容

**Reply by HFrancisla** [reply]
感谢你的回答！ 

我大概清楚.zshrc的内容了 ，我只是想知道有没有办法能够在使用instant prompt的前提下，消除报错以使得均正常工作；
如果将instant prompt注释掉那似乎和

     typeset -g POWERLEVEL10K_INSTANT_PROMPT=off

    * You will not see this error message again.
    * Zsh will start slowly.
   
的效果一致了，这显然不是我希望的。

再表感谢。
